File: Deepspace_ruby/Deepspace/lib/Damage.rb
#encoding:utf-8
require_relative 'DamageToUI'
require 'pp'

module Deepspace

class Damage
  attr_reader :nShields, :nWeapons, :weapons
  
  def initialize(nweapons, nshields,weapons)
    @nShields=nshields
    @nWeapons=nweapons
    @weapons = weapons
  end

  def self.newNumericWeapons(w, s)
    new(w,s,nil)
  end
  
  def self.newSpecificWeapons(wl,s)
    new(-1,s,wl)
  end
  
  def self.newCopy(copy)
    new(copy.nWeapons, copy.nShields, copy.weapons)
  end
  
  def getUIversion
    DamageToUI.new(self)
  end
  
  def adjust(w,s)
    # Primero miramos si es numérico o armas concretas el arma
    if @nWeapons == -1 #Entonces es específico
      weaponTypes = []
      w.each{|elem|
        weaponTypes.push(elem.type)
      }   
      #Calculamos la intersección entre weaponTypes & @weapons
      aux = weaponTypes & @weapons
      for i in 0...aux.length
        k = [weaponTypes.count(aux[i]), @weapons.count(aux[i])].min
        for j in 2..k
          aux.push(aux[i])
        end
      end
      Damage.newSpecificWeapons(aux, [@nShields,s.length].min)      
    else #Entonces es numérico
      Damage.newNumericWeapons([@nWeapons,w.length].min, [@nShields,s.length].min)      
    end    
  end
    
  def discardWeapon(w)
    if @nWeapons == -1 #Entonces es específico
      @weapons.delete(w.type)
    else
      if @nWeapons > 0 
        @nWeapons -= 1
      end
    end     
  end
  
  def discardShieldBooster
    if @nShields > 0
      @nShields -= 1
    end
  end
  
  
  def hasNoEffect
    return @nShields == 0 && (@nWeapons == 0 || @weapons.length == 0)
  end 
  
 def arrayContainsType(w,s)
    for i in 0...w.length
      if w[i].type == s
        return i
      end
    end
    return -1
 end
  
  def to_s
    if nWeapons == -1
      return "\nnShields: #{@nShields} weapons: #{@weapons}"
    end
    return "\nnShields: #{@nShields} nWeapons: #{@nWeapons}"
  end

  private :arrayContainsType
  private_class_method :new
  
end

end

File: Deepspace_java/Deepspace/src/deepspace/Damage.java
package deepspace;
import java.util.ArrayList;
import java.lang.Math;
import java.util.Collections;


/**
 *
 * @author victor
 */
public class Damage {
    private int nShields, nWeapons;
    private ArrayList<WeaponType> weapons;

    Damage( int nWeapons, int nShields) {
        this.nShields = nShields;
        this.nWeapons = nWeapons;
    }

    Damage(ArrayList<WeaponType> weapons, int nShields) {
        this.nShields = nShields;
        this.weapons = weapons;
        this.nWeapons = -1;
    }
    
    Damage(Damage copy) {
        nShields = copy.nShields;
        nWeapons = copy.nWeapons;
        weapons = copy.weapons;        
    }
    
    DamageToUI getUIversion() {
        return new DamageToUI(this);
    }

    public int getNShields() {
        return nShields;
    }

    public int getNWeapons() {
        return nWeapons;
    }

    public ArrayList<WeaponType> getWeapons() {
        return weapons;
    }
      
    
    
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
     //Primero miramos si es numérico o armas concretas el arma
        if (nWeapons == -1) {//Entonces es específico
          ArrayList<WeaponType> aux = new ArrayList<WeaponType>();
          ArrayList<WeaponType> res = new ArrayList<WeaponType>();
          
        for (Weapon elemento:w)
            aux.add(elemento.getType());
               
        WeaponType wtypes [] = WeaponType.values();
        for(int i = 0; i < wtypes.length; i+=1) {
            int min = Math.min(Collections.frequency(aux, wtypes[i]),Collections.frequency(weapons, wtypes[i]));
            for(int j = 0; j < min; j+=1)
                res.add(wtypes[i]);
        } 
        
        return new Damage(res,Math.min(nShields,s.size()));
        } else //Entonces es numérico
           return new Damage(Math.min(nWeapons,w.size()), Math.min(nShields,s.size()));      
      
      }
    
     public void discardWeapon(Weapon w) {
          if(nWeapons == -1) //Entonces es específico
              weapons.remove(w.getType());
           else //Es numérico
              if(nWeapons > 0)
                  nWeapons -= 1;  
         
       }

    @Override
    public String toString() {
        return "Damage{" + "nShields=" + nShields + ", nWeapons=" + nWeapons + ", weapons=" + weapons + '}';
    }
      
   public void discardShieldBooster() {
       if(nShields > 0)
           nShields -= 1;
   }
   
   public boolean hasNoEffect() {
       return nShields == 0 && (nWeapons == 0 || (nWeapons == -1 && weapons.isEmpty()));
   }
   
   private int arrayContainsType(ArrayList<Weapon> w,WeaponType t) {
       for(int i = 0; i < w.size(); i++) 
           if(w.get(i).getType() == t)
               return i;
       
        return -1;       
    }
      
    
}
